### PR TITLE
Move the existing record layer code to use the new record layer design (read side only)

### PR DESCRIPTION
**Pull request by mattcaswell**
This is very much a Work in Progress early-stage preview of what implementing the record layer design described in #17969 looks like for the existing libssl record layer implementations. Most interesting to look at at this stage is the file ssl/record/methods/tlsrecord.c and the header file based on the one in #17969 in ssl/record/recordmethod.h. I intend to break the tlsrecord.c file up into smaller files - because it is becoming too long - but that is future work.

The code as it stands compiles and the tests pass in a default build. It probably doesn't do that for some of the non-default build options. Part of the purpose is to validate the design in #17969 and this work has highlighted a number of things that need to be added to that design.

This work is far from complete. So far I have concentrated only on the "read" (i.e. decryption) side of processing. The "write" side will come later.

Most progressed is the TLSv1.0/TLSv1.1/TLSv1.2 code. This still requires some further work to remove some remaining references to SSL object internals. The intent is not to require the SSL object at all in the record layer method implementations.

The SSLv3.0/TLSv1.3 decryption crypto code has not yet been moved to the new implementation - so this side of things is currently a blend of the old and new code.

The DTLS code has had some code moved across but most of it is still using the old record layer code.

In order to get this to compile and pass the tests in its current intermediate state I've had to implement several "hacks" to get things to work. This includes adding some temporary additional function pointers into the OSSL_RECORD_LAYER structure, as well as passing the SSL object into various function calls which should not be necessary in the final end state. This will all be tidied up in time as I progress through this work.

### CODE REVIEW

**Comment by mattcaswell** [comment]
KTLS will also need some more work. I plan to have a separate KTLS record method implementation - but at the moment its still mixed in with the test of the TLS code.

**Comment by mattcaswell** [comment]
Updated to move the KTLS code to use the new read record layer. Also some significant re-org of the code so far

**Comment by mattcaswell** [comment]
I've now added support to move data received by a record layer object in one epoch into a record layer object in the next epoch. For example, in the case where read_ahead is set and we read a key update message followed by application data all in one go. After the key update is processed, we will be using a new record layer object so we need to move the data we already read, into the new record layer.

This will become particularly important when I start implementing DTLS.

**Comment by mattcaswell** [comment]
The CIs are now passing for this PR except for one - which is in the external tests. Specifically the oqsprovider test:

````
patching file test/ssltestlib.c
Hunk #9 succeeded at 835 (offset -1 lines).
Hunk #10 succeeded at 872 (offset -1 lines).
Hunk #11 succeeded at 964 (offset 17 lines).
Hunk #12 succeeded at 998 (offset 33 lines).
Hunk #13 FAILED at 981.
Hunk #14 succeeded at 1036 (offset 33 lines).
Hunk #15 succeeded at 1068 (offset 38 lines).
1 out of 15 hunks FAILED -- saving rejects to file test/ssltestlib.c.rej
````

The problem seems to be that this PR amends test/ssltestlib.c - but the oqsprovider external test is patching that file. The conflict is causing the patch and hence the whole test to fail.

This seems like a chicken-and-egg problem. We can't fix the oqsprovider yet to work with this PR, because it would (presumably) start failing on master. However if this PR goes in with that test failing, then its going to start failing everywhere. Probably I will have to disable the test temporarily in this PR, and then we fix it afterwards once its gone in. There's still a lot of work to do on this PR - so it will be a while before we get to that point anyway.

Ping @baentsch for a heads up. Nothing to do at this stage, but we'll have to fix this once this PR goes in.

**Comment by baentsch** [comment]
@mattcaswell You may want to run CI again: The `oqsprovider` test should now pass again (assuming this PR doesn't change the "helpers" test API too drastically).

**Comment by mattcaswell** [comment]
@baentsch Thanks! I've updated the oqs-provider submodule in this PR to the latest commit in order to pick up your latest changes. I also had to modify our oqsprovider.sh script to remove a reference to the preptests.sh script that you removed as part of your fix. Hopefully I did it right. Lets see what the CIs make of it now.

**Comment by mattcaswell** [comment]
Updated the read record layer to clean away all the remaining references to the SSL object. Just DTLS left to look at on the read record layer side of things.

**Comment by mattcaswell** [comment]
CIs now all green except one: the GOST external test. Not too sure how to debug that one - any hints @beldmit?

**Comment by beldmit** [comment]
```
9/15 Test  #9: TLS-with-engine ..................Subprocess aborted***Exception:   0.73 sec
262
40070A504C7F0000:error:0A000119:SSL routines:tls_get_more_records:decryption failed or bad record mac:ssl/record/methods/tls_common.c:766:
263
40070A504C7F0000:error:0A000139:SSL routines::record layer failure:ssl/record/rec_layer_s3.c:1197:
264
/home/runner/work/openssl/openssl/gost-engine/test_tls.c:222: OpenSSL internal error: SSL_accept
```

**Comment by beldmit** [comment]
Good news it's C code...

**Comment by mattcaswell** [comment]
That much I figured out :-) What I'm not sure on is how to set that test up so I can run it through a debugger...

**Comment by beldmit** [comment]
Try this from the engine build dir:
```
LD_LIBRARY_PATH=../openssl OPENSSL_CONF=./engine.conf ./bin/test_tls
```

engine.conf:
```
openssl_conf = openssl_def
[openssl_def]
engines = engine_section

[engine_section]
gost = gost_section

[gost_section]
engine_id = gost
dynamic_path = /home/dbelyavs/upstream/engine/bin/gost.so
default_algorithms = ALL
CRYPT_PARAMS = id-Gost28147-89-CryptoPro-A-ParamSet

[req]
distinguished_name = req_dn
[ req_dn ]
C = RU
L = Moscow
CN=Test srv gost2012_256:XA
O=Cryptocom
OU=OpenSSL CA
emailAddress = openssl@cryptocom.ru

```

**Comment by mattcaswell** [comment]
Great - thanks. Will try that.

**Comment by beldmit** [comment]
Feel free to ping me directly, I'm able to have a google meet if necessary

**Comment by mattcaswell** [comment]
> Feel free to ping me directly, I'm able to have a google meet if necessary

Will do - thanks. Let's see how far I get first!

**Comment by mattcaswell** [comment]
Thanks for tips @beldmit. I've now isolated the final problem and CIs are green!!

I still have some outstanding TODOs on this PR that I am working through.

**Comment by mattcaswell** [comment]
Finally I think this PR is ready for review!!

I have taken it out of draft. This PR includes the commits from #18596 and #18868 which are being reviewed separately. As those PRs get merged I will rebase this one so that they disappear from here.

This PR only handles the read side of the record layer. A separate PR will build on this work to implement the write side.

Please take a look!

**Comment by mattcaswell** [comment]
I've rebased this on top of commit ce602bb0 from the master branch. That includes the commits from #18596, so they no longer appear here. I've also removed the commits from #18868 from this PR. They are not strictly necessary for this PR. It is mainly useful because it adds a test for some DTLS specific record layer behaviour. However this PR can still work without that.

Therefore all external dependencies are removed from this PR and it can be reviewed as is.

There are still a significant number of conflicts between this PR and the head of master. Those are due to changes merged in commit 38b051a1 as a result of the SSL object refactor (PR #18612). I will work on resolving those conflicts separately - but I think reviewing of this PR can proceed in the meantime.

**Comment by mattcaswell** [comment]
This PR has now been rebased to the head of master, so now incorporates all of the SSL object refactor changes.

File: ssl/record/recordmethod.h [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+     * the record layer has failed because it is unsuitable, but an alternative
+     * record layer can be tried instead.
+     */
+
+    /*
+     * If we eventually make this fetchable then we will need to use something
+     * other than EVP_CIPHER. Also mactype would not be a NID, but a string. For
+     * now though, this works.
+     */
+    int (*new_record_layer)(OSSL_LIB_CTX *libctx,
```
Comment (t8m) [inline ssl/record/recordmethod.h:124]: Food for thought: Would it make sense to make some of the huge number of the parameters an OSSL_PARAM settings params?

Would it allow to actually use the RL API also for the QUIC RL implementation?

File: test/sslapitest.c [hunk 2]
```diff
@@ -9395,7 +9397,7 @@ static int test_sni_tls13(void)
         goto end;
     /* Require TLSv1.3 as a minimum */
     if (!TEST_true(create_ssl_ctx_pair(libctx, TLS_server_method(),
-                                       TLS_client_method(), TLS1_3_VERSION, 0,
+                                       TLS_client_method(), 0, 0,
```
Comment (t8m) [inline test/sslapitest.c:9400]: Why this change?

File: ssl/record/ssl3_record.c [hunk 3]
```diff
@@ -987,66 +261,51 @@ int tls1_enc(SSL_CONNECTION *s, SSL3_RECORD *recs, size_t n_recs, int sending,
...
-        if (EVP_MD_CTX_get0_md(s->write_hash)) {
-            int n = EVP_MD_CTX_get_size(s->write_hash);
-            if (!ossl_assert(n >= 0)) {
-                SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
-                return 0;
-            }
+    assert(sending);
+    if (EVP_MD_CTX_get0_md(s->write_hash)) {
+        int n = EVP_MD_CTX_get_size(s->write_hash);
+        if (!ossl_assert(n >= 0)) {
```
Comment (t8m) [inline ssl/record/ssl3_record.c:274]: Missing empty line

**Review by t8m — COMMENTED** [review]
By no means complete review. Just a few comments.

This is great refactoring work, Matt!

File: ssl/record/recordmethod.h (hunk 1 above)
Comment (mattcaswell) [reply ssl/record/recordmethod.h:124]: Maybe, although I'd prefer to keep parameters that are fundamental to the operation of the record layer as direct parameters. We also know that parsing parameters is not the most performance efficient thing to do, so I'd rather not do *everything* as OSSL_PARAM just because we can. OTOH, I fully expect to revise this when we make this interface public at some point in the future (and at that point it will probably be fetchable). Nothing is set in stone while it is an internal API only.

File: test/sslapitest.c (hunk 2 above)
Comment (mattcaswell) [reply test/sslapitest.c:9400]: I have no idea!! I don't remember changing that and it doesn't look like a correct change either. Although in practice it probably doesn't make a difference because this code won't be compiled unless we have usable TLSv1.3 and we will always use it by default. I'll fix it.

**Comment by mattcaswell** [comment]
I've fixed or otherwise responded to all feedback items above. I've also rebased since there were some conflicts.

File: ssl/record/methods/tls_common.c [hunk 4]
```diff
@@ -0,0 +1,1408 @@
...
+                 * to SSL2_VERSION. This way validate_record_header() can know
+                 * what format the record was in based on the version.
+                 */
+                if (thisrr->rec_version == SSL2_VERSION) {
+                    RLAYERfatal(rl, SSL_AD_PROTOCOL_VERSION,
+                                SSL_R_WRONG_VERSION_NUMBER);
+                    return OSSL_RECORD_RETURN_FATAL;
+                }
+
+                rl->msg_callback(0, version, SSL3_RT_HEADER, p, 5, rl->cbarg);
```
Comment (hlandau) [inline ssl/record/methods/tls_common.c:522]: I guess this is always non-NULL?

**Review by hlandau — APPROVED** [review]
This was arduous to review... my familiarity with this code is limited, but I've gone through it commit by commit and think it should be merged.

File: ssl/record/methods/tls_common.c (hunk 4 above)
Comment (mattcaswell) [reply ssl/record/methods/tls_common.c]: It always is (libssl will always ensure this) - but strictly speaking the record layer should check this itself (which it doesn't). I've pushed a commit to do that.

**Review by t8m — APPROVED** [review]
Great work! I spent more than two afternoons going through this and was not really able to find any issue in this huge PR.

**Comment by t8m** [comment]
@hlandau can you please reconfirm as there was a small commit added after your review?

**Review by hlandau — APPROVED** [review]

**Comment by hlandau** [comment]
Still good.

**Comment by mattcaswell** [comment]
Pushed!! Thanks @t8m and @hlandau for your time spent reviewing this!
